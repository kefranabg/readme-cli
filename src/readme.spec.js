const fs = require('fs')
const ora = require('ora')

jest.mock('ora')

const { writeReadme, buildReadmeContent, README_PATH } = require('./readme')

describe('readme', () => {
  const succeed = jest.fn()
  const fail = jest.fn()

  ora.mockReturnValue({
    start: () => ({
      succeed,
      fail
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('writeReadme', () => {
    it('should call ora with correct parameters in success case', async () => {
      const readmeContent = 'content'
      fs.writeFile = jest.fn((path, content, cb) => cb(null, 'done'))

      await writeReadme(readmeContent)

      expect(ora).toHaveBeenCalledTimes(1)
      expect(ora).toHaveBeenCalledWith('Creating README')
      expect(succeed).toHaveBeenCalledTimes(1)
      expect(succeed).toHaveBeenCalledWith('README created')
    })

    it('should call ora with correct parameters in fail case', async () => {
      const readmeContent = 'content'
      fs.writeFile = jest.fn(() => {
        throw new Error('error')
      })

      try {
        await writeReadme(readmeContent)
        // eslint-disable-next-line no-empty
      } catch (err) {}

      expect(ora).toHaveBeenCalledTimes(1)
      expect(ora).toHaveBeenCalledWith('Creating README')
      expect(fail).toHaveBeenCalledTimes(1)
      expect(fail).toHaveBeenCalledWith('README creation fail')
    })

    it('should call writeFile with correct parameters', async () => {
      const readmeContent = 'content'
      fs.writeFile = jest.fn((path, content, cb) => cb(null, 'done'))

      await writeReadme(readmeContent)

      expect(fs.writeFile).toHaveBeenCalledTimes(1)
      expect(fs.writeFile.mock.calls[0][0]).toBe(README_PATH)
      expect(fs.writeFile.mock.calls[0][1]).toBe(readmeContent)
    })
  })

  describe('buildReadmeContent', () => {
    const templateName = 'default'
    const context = {
      isGithubRepos: true,
      repositoryUrl: 'https://github.com/kefranabg/readme-md-generator',
      projectPrerequisites: [
        { name: 'npm', value: '>=5.5.0' },
        { name: 'node', value: '>= 9.3.0' }
      ],
      projectName: 'readme-md-generator',
      projectVersion: '0.1.3',
      projectDescription:
        'Generates beautiful README files from git config & package.json infos',
      projectDocumentationUrl:
        'https://github.com/kefranabg/readme-md-generator#readme',
      projectHomepage:
        'https://github.com/kefranabg/readme-md-generator#readme',
      authorName: 'Franck Abgrall',
      authorGithubUsername: 'kefranabg',
      authorTwitterUsername: 'FranckAbgrall',
      licenseName: 'MIT',
      licenseUrl:
        'https://github.com/kefranabg/readme-md-generator/blob/master/LICENSE',
      contributingUrl:
        'https://github.com/kefranabg/readme-md-generator/issues',
      installCommand: 'npm install',
      usage: 'npm start',
      testCommand: 'npm run test'
    }

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should call ora with correct parameters in success case', async () => {
      await buildReadmeContent(context, templateName)

      expect(ora).toHaveBeenCalledTimes(1)
      expect(ora).toHaveBeenCalledWith('Loading README template')
      expect(succeed).toHaveBeenCalledTimes(1)
      expect(succeed).toHaveBeenCalledWith('README template loaded')
    })

    it('should return readme template content', async () => {
      const result = await buildReadmeContent(context, templateName)

      expect(result).toMatchSnapshot()
    })

    it('should call ora with correct parameters in fail case', async () => {
      fs.readFile = jest.fn(() => {
        throw new Error('error')
      })

      try {
        await buildReadmeContent(context, templateName)
        // eslint-disable-next-line no-empty
      } catch (err) {}

      expect(ora).toHaveBeenCalledTimes(1)
      expect(ora).toHaveBeenCalledWith('Loading README template')
      expect(fail).toHaveBeenCalledTimes(1)
      expect(fail).toHaveBeenCalledWith('README template loading fail')
    })
  })
})
