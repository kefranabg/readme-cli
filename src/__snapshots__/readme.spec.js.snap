// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readme buildReadmeContent should return readme template content 1`] = `
"<h1 align=\\"center\\">Welcome to readme-md-generator 👋</h1>
<p>
  <img src=\\"https://img.shields.io/badge/version-0.1.3-blue.svg?cacheSeconds=2592000\\" />
  <img src=\\"https://img.shields.io/badge/npm-%3E%3D5.5.0-blue.svg\\" />
  <img src=\\"https://img.shields.io/badge/node-%3E%3D%209.3.0-blue.svg\\" />
  <a href=\\"https://github.com/kefranabg/readme-md-generator#readme\\">
    <img alt=\\"Documentation\\" src=\\"https://img.shields.io/badge/documentation-yes-brightgreen.svg\\" target=\\"_blank\\" />
  </a>
  <a href=\\"https://github.com/kefranabg/readme-md-generator/graphs/commit-activity\\">
    <img alt=\\"Maintenance\\" src=\\"https://img.shields.io/badge/Maintained%3F-yes-green.svg\\" target=\\"_blank\\" />
  </a>
  <a href=\\"https://github.com/kefranabg/readme-md-generator/blob/master/LICENSE\\">
    <img alt=\\"License: MIT\\" src=\\"https://img.shields.io/badge/License-MIT-yellow.svg\\" target=\\"_blank\\" />
  </a>
  <a href=\\"https://twitter.com/FranckAbgrall\\">
    <img alt=\\"Twitter: FranckAbgrall\\" src=\\"https://img.shields.io/twitter/follow/FranckAbgrall.svg?style=social\\" target=\\"_blank\\" />
  </a>
</p>

> Generates beautiful README files from git config &amp; package.json infos

### 🏠 [Homepage](https://github.com/kefranabg/readme-md-generator#readme)

## Prerequisites

- npm &gt;=5.5.0
- node &gt;= 9.3.0

## Install

\`\`\`sh
npm install
\`\`\`

## Usage

\`\`\`sh
npm start
\`\`\`

## Run tests

\`\`\`sh
npm run test
\`\`\`

## Author

👤 **Franck Abgrall**

* Twitter: [@FranckAbgrall](https://twitter.com/FranckAbgrall)
* Github: [@kefranabg](https://github.com/kefranabg)

## 🤝 Contributing

Contributions, issues and feature requests are welcome!<br />Feel free to check [issues page](https://github.com/kefranabg/readme-md-generator/issues).

## Show your support

Give a ⭐️ if this project helped you!

## 📝 License

Copyright © 2019 [Franck Abgrall](https://github.com/kefranabg).<br />
This project is [MIT](https://github.com/kefranabg/readme-md-generator/blob/master/LICENSE) licensed.

***
_This README was generated with ❤️ by [readme-md-generator](https://github.com/kefranabg/readme-md-generator)_"
`;
